e2e/playwright/conftest.py

```diff
@@ -0,0 +1,365 @@
+
+import os
+import re
+import shutil
+import socket
+import subprocess
+import sys
+import time
+from io import BytesIO
+from pathlib import Path
+from random import randint
+from tempfile import TemporaryFile
+from typing import Generator, List, Protocol
+
+import pytest
+import requests
+from PIL import Image
+from playwright.sync_api import Page
+from pytest import FixtureRequest
+
+
+class AsyncSubprocess:
+    """A context manager. Wraps subprocess.Popen to capture output safely."""
+
+    def __init__(self, args, cwd=None, env=None):
+        self.args = args
+        self.cwd = cwd
+        self.env = env
+        self._proc = None
+        self._stdout_file = None
+
+    def terminate(self):
+        """Terminate the process and return its stdout/stderr in a string."""
+        # Terminate the process
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc.wait()
+            self._proc = None
+
+        # Read the stdout file and close it
+        stdout = None
+        if self._stdout_file is not None:
+            self._stdout_file.seek(0)
+            stdout = self._stdout_file.read()
+            self._stdout_file.close()
+            self._stdout_file = None
+
+        return stdout
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def start(self):
+        # Start the process and capture its stdout/stderr output to a temp
+        # file. We do this instead of using subprocess.PIPE (which causes the
+        # Popen object to capture the output to its own internal buffer),
+        # because large amounts of output can cause it to deadlock.
+        self._stdout_file = TemporaryFile("w+")
+        print(f"Running: {self.args}")
+        self._proc = subprocess.Popen(
+            self.args,
+            cwd=self.cwd,
+            stdout=self._stdout_file,
+            stderr=subprocess.STDOUT,
+            text=True,
+            env={**os.environ.copy(), **self.env} if self.env else None,
+        )
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self._proc is not None:
+            self._proc.terminate()
+            self._proc = None
+        if self._stdout_file is not None:
+            self._stdout_file.close()
+            self._stdout_file = None
+
+
+def is_port_available(port: int, host: str) -> bool:
+    """Check if a port is available on the given host."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+        return sock.connect_ex((host, port)) != 0
+
+
+def find_available_port(
+    min_port: int = 10000,
+    max_port: int = 65535,
+    max_tries: int = 50,
+    host: str = "localhost",
+) -> int:
+    """Find an available port on the given host."""
+    for _ in range(max_tries):
+        selected_port = randint(min_port, max_port)
+        if is_port_available(selected_port, host):
+            return selected_port
+    raise RuntimeError("Unable to find an available port.")
```

Comment:
```suggestion
def find_available_port(host: str = "localhost") -> int:
    """Find an available port on the given host."""
    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
        s.bind((host, 0))
        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        return s.getsockname()[1]
```
We can bind to port 0 to get a random available port. See: https://stackoverflow.com/questions/1075399/how-to-bind-to-any-available-port

Reply:
Oh, I didn't know about `0` for this. Thanks for pointing this out. This is indeed a much better solution here!